CameraX-MLKit/app/src/main/java/com/example/camerax_mlkit/QrCodeDrawable.kt

```diff
@@ -0,0 +1,93 @@
+package com.example.camerax_mlkit
+
+import android.content.Intent
+import android.graphics.*
+import android.graphics.drawable.Drawable
+import android.net.Uri
+import android.view.MotionEvent
+import android.view.View
+import com.google.mlkit.vision.barcode.common.Barcode
+
+
+class QrCodeDrawable(barcode: Barcode) : Drawable() {
+    private val boundingRectPaint = Paint().apply {
+        style = Paint.Style.STROKE
+        color = Color.YELLOW
+        strokeWidth = 5F
+        alpha = 200
+    }
+
+    private val contentRectPaint = Paint().apply {
+        style = Paint.Style.FILL
+        color = Color.YELLOW
+        alpha = 255
+    }
+
+    private val contentTextPaint = Paint().apply {
+        color = Color.DKGRAY
+        alpha = 255
+        textSize = 36F
+    }
+
+    private var boundingRect: Rect = barcode.boundingBox!!
+    private val contentPadding = 25
+
+    private var qrContent: String = ""
+    private var textWidth: Int = 0
+    var qrCodeTouchCallback = { v: View, e: MotionEvent -> false} //no-op
```

Comment:
nit: Having a UI event callback in a Drawable is unintuitive. Can we move this to the `MainActivity`?

Reply:
The original reason I didn't put this in MainActivity was to avoid duplicating the logic for checking the QR Code data type. Still, I agree it doesn't belong in the Drawable. So now I've extracted this logic out into a QrCodeViewModel.